design/Events_design.md

```diff
@@ -0,0 +1,300 @@
+- the visibility rules for events.
+- technical implementation details.
+
+## Ethereum Events
+
+To help dApp developers design applications with a good UX, the ethereum developers invented the concept of "events" or "logs", which
+are pieces of information emitted from smart contracts, which can be streamed in real time to external applications that
+subscribe to them.
+
+To better understand the anatomy of events, read this [blog](https://medium.com/mycrypto/understanding-event-logs-on-the-ethereum-blockchain-f4ae7ba50378)
+
+### Smart contracts
+
+This is how an event is declared in a smart contract.
+
+```
+event Transfer(address indexed from, address indexed to, uint256 value);
+```
+
+And this is how it is emitted.
+```
+emit Transfer(from, to, amount);
+```
+
+### Consuming events
+
+A web app can subscribe to events by doing something like:
+
+```
+var subscription = web3.eth.subscribe('logs', {
+    address: '0x123456..', 
+    topics: ['0x12345...']
+}, function(error, result){
+    if (!error)
+        console.log(result);
+});
+- address - String|Array: An address or a list of addresses to only get logs from particular account(s).
+- topics - Array: An array of values which must each appear in the log entries. The order is important, if you want to leave topics out use null, e.g. [null, '0x00...']. You can also pass another array for each topic with options for that topic e.g. [null, ['option1', 'option2']]
+```
+
+A web app might request:
+
+- stream all "transfer" events where the 'to' field is my address
+- stream all "transfer" events from the USDC contract
+
+Apps can also request historic events starting from any block.
+
+### How events work
+
+`Note: this section might not be 100% accurate.`
+
+The query made on the UI side is transformed in a server-side query on the node, and there is some logic after a tx is
+executed and events are emmitted, to match them against the filters requested by users and distribute them to the
+requester.
+
+Todo: Can you clarify in the doc whether the events are discarded after being distributed, or are stored for future
+subscribers?
+
+*Note that there is no constraint on data access, since all data is public.*
+
+## Obscuro Design
+
+In Obscuro, we aim to maintain the same building blocks that are found in Ethereum: events and subscriptions, and will try
+to implement the privacy concerns with as little disruption as possible.
+
+### Event visibility rules
+
+There are a couple of cases that must be considered in order to decide whether Alice is entitled to view an event:
+
+1. The event was emitted by a smart contract as a result of executing a Tx sent by Alice.
+2. The event was emitted as result of a Tx sent by Bob which is relevant to Alice. (See below for a definition of
+   relevancy.)
+3. The event was emitted as result of a Tx sent by Bob which is not relevant to Alice.
+
+*Note that we assume that all events emitted during a transaction signed by Alice, will be visible to Alice.*  
```

Comment:
Can you clarify which of the three cases above we support? Based on this sentence, we support (1). Do we also support (2) and (3)?

Reply:
Ah, you cover this below. I think maybe get rid of this sentence here, since you cover below who gets to see what.

Reply:
yeah, it's out of place here.